internal/freelist/freelist_test.go

```diff
@@ -34,6 +38,56 @@ func TestFreelist_free_overflow(t *testing.T) {
 	}
 }
 
+// Ensure that double freeing a page is causing a panic
+func TestFreelist_free_double_free_panics(t *testing.T) {
+	f := newTestFreelist()
+	f.Free(100, common.NewPage(12, 0, 0, 3))
+	require.Panics(t, func() {
+		f.Free(100, common.NewPage(12, 0, 0, 3))
+	})
+}
+
+// Ensure that attempting to free the meta page panics
+func TestFreelist_free_meta_panics(t *testing.T) {
+	f := newTestFreelist()
+	require.Panics(t, func() {
+		f.Free(100, common.NewPage(0, 0, 0, 0))
+	})
+	require.Panics(t, func() {
+		f.Free(100, common.NewPage(1, 0, 0, 0))
+	})
+}
+
+func TestFreelist_free_freelist(t *testing.T) {
+	f := newTestFreelist()
+	f.Free(100, common.NewPage(12, common.FreelistPageFlag, 0, 0))
+	pp := f.pendingPageIds()[100]
+	require.Equal(t, []common.Pgid{12}, pp.ids)
+	require.Equal(t, []common.Txid{0}, pp.alloctx)
+}
+
+func TestFreelist_free_freelist_alloctx(t *testing.T) {
+	f := newTestFreelist()
+	f.Free(100, common.NewPage(12, common.FreelistPageFlag, 0, 0))
+	f.Rollback(100)
+	require.Empty(t, f.freePageIds())
+	require.Empty(t, f.pendingPageIds())
+	require.False(t, f.Freed(12))
+
+	// we still hold an allotx reference to page 12 through txid 99 - let's try to free it again
```

Comment:
I've removed this comment. txid 99 was because we had this little hack:
https://github.com/etcd-io/bbolt/commit/d72e6bf8c17eca17a619ffc1648a40e99b93c151#diff-d722be595713f8fc49d086ab1968ff45f223c2b1f99bc0f7ae65fbdde3132c40L70-L73